doc/speedtest.3
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "App::SpeedTest 3"
.TH App::SpeedTest 3 2025-03-14 "perl v5.40.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
App::SpeedTest \- Command\-line interface to speedtest.net
.SH SYNOPSIS
.IX Header "SYNOPSIS"
.Vb 1
\& $ speedtest [ \-\-no\-geo | \-\-country=NL ] [ \-\-list | \-\-ping ] [ options ]
\&
\& $ speedtest \-\-list
\& $ speedtest \-\-ping \-\-country=BE
\& $ speedtest
\& $ speedtest \-s 4358
\& $ speedtest \-\-url=http://ookla.extraip.net
\& $ speedtest \-q \-\-no\-download
\& $ speedtest \-Q \-\-no\-upload
.Ve
.SH DESCRIPTION
.IX Header "DESCRIPTION"
The provided perl script is a command\-line interface to the
speedtest.net <http://www.speedtest.net/> infrastructure so that
flash is not required
.PP
It was written to feature the functionality that speedtest.net offers
without the overhead of flash or java and the need of a browser.
.SH Raison\-d\*(Aqêtre
.IX Header "Raison-d'être"
The tool is there to give you a quick indication of the achievable
throughput of your current network. That can drop dramatically if
you are behind (several) firewalls or badly configured networks (or
network parts like switches, hubs and routers).
.PP
It was inspired by speedtest\-cli <https://github.com/sivel/speedtest-cli>,
a project written in python. But I neither like python, nor did I like the
default behavior of that script. I also think it does not take the right
decisions in choosing the server based on distance instead of speed. That
\&\fBdoes\fR matter if one has fiber lines. I prefer speed over distance.
.SH "Command\-line Arguments"
.IX Xref "CLIA"
.IX Header "Command-line Arguments"
.IP "\-? | \-\-help" 2
.IX Xref "-? --help"
.IX Item "-? | --help"
Show all available options and then exit.
.IP "\-V | \-\-version" 2
.IX Xref "-V --version"
.IX Item "-V | --version"
Show program version and exit.
.IP \-\-man 2
.IX Xref "--man"
.IX Item "--man"
Show the builtin manual using \f(CW\*(C`pod2man\*(C'\fR and \f(CW\*(C`nroff\*(C'\fR.
.IP \-\-info 2
.IX Xref "--info"
.IX Item "--info"
Show the builtin manual using \f(CW\*(C`pod2text\*(C'\fR.
.IP "\-v[#] | \-\-verbose[=#]" 2
.IX Xref "-v --version"
.IX Item "-v[#] | --verbose[=#]"
Set verbose level. Default value is 1. A plain \-v without value will set
the level to 2.
.IP \-\-simple 2
.IX Xref "--simple"
.IX Item "--simple"
An alias for \f(CW\*(C`\-v0\*(C'\fR
.IP \-\-all 2
.IX Xref "--all"
.IX Item "--all"
No (default) filtering on available servers. Useful when finding servers
outside of the country of your own location.
.IP "\-g | \-\-geo" 2
.IX Xref "-g --geo"
.IX Item "-g | --geo"
Use GEO\-IP service to find the country your ISP is located. The default
is true. If disable (\f(CW\*(C`\-\-no\-geo\*(C'\fR), the server to use will be based on
distance instead of on latency.
.IP "\-cXX | \-\-cc=XX | \-\-country=XX" 2
.IX Xref "-c --cc --country"
.IX Item "-cXX | --cc=XX | --country=XX"
Pass the ISO country code to select the servers
.Sp
.Vb 3
\& $ speedtest \-c NL ...
\& $ speedtest \-\-cc=B ...
\& $ speedtest \-\-country=D ...
.Ve
.IP \-\-list\-cc 2
.IX Xref "--list-cc"
.IX Item "--list-cc"
Fetch the server list and then show the list of countries the servers are
located with their country code and server count
.Sp
.Vb 5
\& $ speedtest \-\-list\-cc
\& AD Andorra                             1
\& AE United Arab Emirates                4
\& :
\& ZW Zimbabwe                            6
.Ve
.Sp
You can then use that code to list the servers in the chosen country, as
described below.
.IP "\-l | \-\-list" 2
.IX Xref "-l --list"
.IX Item "-l | --list"
This option will show all servers in the selection with the distance in
kilometers to the server.
.Sp
.Vb 11
\& $ speedtest \-\-list \-\-country=IS
\&   1: 10661 \- Tengir hf              Akureyri    1980.02 km
\&   2: 21605 \- Premis ehf             Reykjavík   2039.16 km
\&   3:  3684 \- Nova                   Reykjavik   2039.16 km
\&   4:  6471 \- Gagnaveita Reykjavikur Reykjavik   2039.16 km
\&   5: 10650 \- Nova VIP               Reykjavik   2039.16 km
\&   6: 16148 \- Hringidan              Reykjavik   2039.16 km
\&   7:  4818 \- Siminn                 Reykjavik   2039.16 km
\&   8: 17455 \- Hringdu                Reykjavík   2039.16 km
\&   9:  4141 \- Vodafone               Reykjavík   2039.16 km
\&  10:  3644 \- Snerpa                 Isafjordur  2192.27 km
.Ve
.IP "\-p | \-\-ping | \-\-ping=40" 2
.IX Xref "-p --ping"
.IX Item "-p | --ping | --ping=40"
Show a list of servers in the selection with their latency in ms.
Be very patient if running this with "\-\-all".
.Sp
.Vb 10
\& $ speedtest \-\-ping \-\-cc=BE
\&   1:  4320 \- EDPnet               Sint\-Niklaas     148.06 km      52 ms
\&   2: 12627 \- Proximus             Brussels         173.04 km      55 ms
\&   3: 10986 \- Proximus             Schaarbeek       170.54 km      55 ms
\&   4: 15212 \- Telenet BVBA/SPRL    Mechelen         133.89 km      57 ms
\&   5: 29238 \- Arcadiz              DIEGEM           166.33 km      58 ms
\&   6:  5151 \- Combell              Brussels         173.04 km      59 ms
\&   7: 26887 \- Arxus NV             Brussels         173.04 km      64 ms
\&   8:  4812 \- Universite Catholiq… Louvain\-La\-Neuv  186.87 km      70 ms
\&   9:  2848 \- Cu.be Solutions      Diegem           166.33 km      75 ms
\&  10: 12306 \- VOO                  Liège            186.26 km      80 ms
\&  11: 24261 \- Une Nouvelle Ville…  Charleroi        217.48 km     147 ms
\&  12: 30594 \- Orange Belgium       Evere            169.29 km     150 ms
.Ve
.Sp
If a server does not respond, a very high latency is used as default.
.Sp
This option only shows the 40 nearest servers. The number can be changed
as optional argument.
.Sp
.Vb 5
\& $ speedtest \-\-cc=BE \-\-ping=4
\&   1:  4320 \- EDPnet               Sint\-Niklaas     148.06 km      53 ms
\&   2: 29238 \- Arcadiz              DIEGEM           166.33 km      57 ms
\&   3: 15212 \- Telenet BVBA/SPRL    Mechelen         133.89 km      62 ms
\&   4:  2848 \- Cu.be Solutions      Diegem           166.33 km      76 ms
.Ve
.IP "\-1 | \-\-one\-line" 2
.IX Xref "-1 --ono-line"
.IX Item "-1 | --one-line"
Generate a very short report easy to paste in e.g. IRC channels.
.Sp
.Vb 2
\& $ speedtest \-1Qv0
\& DL:   40.721 Mbit/s, UL:   30.307 Mbit/s
.Ve
.IP "\-B | \-\-bytes" 2
.IX Xref "-B --bytes"
.IX Item "-B | --bytes"
Report throughput in Mbyte/s instead of Mbit/s
.IP "\-C | \-\-csv" 2
.IX Xref "-C --csv"
.IX Item "-C | --csv"
Generate the measurements in CSV format. The data can be collected in
a file (by a cron job) to be able to follow internet speed over time.
.Sp
The reported fields are
.Sp
.Vb 8
\& \- A timestam (the time the tests are finished)
\& \- The server ID
\& \- The latency in ms
\& \- The number of tests executed in this measurement
\& \- The direction of the test (D = Down, U = Up)
\& \- The measure avarage speed in Mbit/s
\& \- The minimum speed measured in one of the test in Mbit/s
\& \- The maximum speed measured in one of the test in Mbit/s
\&
\& $ speedtest \-Cs4358
\& "2015\-01\-24 17:15:09",4358,63.97,40,D,93.45,30.39,136.93
\& "2015\-01\-24 17:15:14",4358,63.97,40,U,92.67,31.10,143.06
.Ve
.IP "\-U | \-\-skip\-undef" 2
.IX Xref "-U --skip-undef"
.IX Item "-U | --skip-undef"
Skip reporting measurements that have no speed recordings at all.
The default is to report these as \f(CW0.00\fR .. \f(CW999999999.999\fR.
.IP "\-P | \-\-prtg" 2
.IX Xref "-P --prtg"
.IX Item "-P | --prtg"
Generate the measurements in XML suited for PRTG
.Sp
.Vb 10
\& $ speedtest \-P
\& <?xml version="1.0" encoding="UTF\-8" ?>
\& <prtg>
\&   <text>Testing from My ISP (10.20.30.40)</text>
\&   <result>
\&     <channel>Ping</channel>
\&     <customUnit>ms</customUnit>
\&     <float>1</float>
\&     <value>56.40</value>
\&     </result>
\&   <result>
\&     <channel>Download</channel>
\&     <customUnit>Mbit/s</customUnit>
\&     <float>1</float>
\&     <value>38.34</value>
\&     </result>
\&   <result>
\&     <channel>Upload</channel>
\&     <customUnit>Mbit/s</customUnit>
\&     <float>1</float>
\&     <value>35.89</value>
\&     </result>
\&   </prtg>
.Ve
.IP \-\-url[=XXX] 2
.IX Xref "--url"
.IX Item "--url[=XXX]"
With no value, show server url in list
.Sp
With value, use specific server url: do not scan available servers
.IP \-\-ip 2
.IX Xref "--ip"
.IX Item "--ip"
Show IP for server
.IP "\-T[#] | \-\-try[=#]" 2
.IX Xref "-T --try"
.IX Item "-T[#] | --try[=#]"
Use the top # (based on lowest latency or shortest distance) from the list
to do all required tests.
.Sp
.Vb 2
\& $ speedtest \-T3 \-c NL \-Q2
\& Testing for 80.x.y.z : XS4ALL Internet BV (NL)
\&
\& Using 13218:  26.52 km      25 ms XS4ALL Internet BV
\& Test download ..                                      Download     31.807 Mbit/s
\& Test upload   ..                                      Upload       86.587 Mbit/s
\&
\& Using 15850:  26.09 km      25 ms QTS Data Centers
\& Test download ..                                      Download     80.763 Mbit/s
\& Test upload   ..                                      Upload       77.122 Mbit/s
\&
\& Using 11365:  26.09 km      27 ms Vancis
\& Test download ..                                      Download    106.022 Mbit/s
\& Test upload   ..                                      Upload       82.891 Mbit/s
\&
\& Rank 01: Server:  11365   26.09 km      27 ms,  DL:  106.022 UL:   82.891
\& Rank 02: Server:  15850   26.09 km      25 ms,  DL:   80.763 UL:   77.122
\& Rank 03: Server:  13218   26.52 km      25 ms,  DL:   31.807 UL:   86.587
\&
\& $ speedtest \-1v0 \-T5
\& DL:  200.014 Mbit/s, UL:  159.347 Mbit/s, SRV: 13218
\& DL:  203.599 Mbit/s, UL:  166.247 Mbit/s, SRV: 15850

ChangeLog
0.31	2025-03-17
    * 

0.30	2025-01-06
    * Add SECURITY.md
    * Fix document encoding

0.29	2024-06-18
    * Make requirements more specific based on known CVE's
    * Work around groff-1.24
    * It is 2024
    * Tested with perl-5.40

0.28	2022-01-01
    * It is 2021 - No changes in 2021
    * It is 2022

0.27	2020-12-23
    * Some man pages are not the same as other man pages

0.26	2020-12-17
    * Fix META to include bugtracker

0.25	2020-04-14
    * Require List::Util for consistency
    * Add static docs for github
    * Allow multiple -s ID

0.24	2020-04-10
    * Switch to Pod::Text when nroff is not available

0.23	2020-04-09
    * Move pod tests to xt
    * Add simple tests for basic CLI
    * Make ping-list-length settable
    * Do not truncate server list when a specific server is selected
    * It's 2020

0.22	2019-09-02
    * Add --list-cc
    * Add cpanfile
    * Add -U / --skip-undef
    * Provide cpanfile
    * Add help and options to examples/speedtest-graph.pl
    * Fix warning when using --no-upload or --no-download (issue#13)

0.21	2019-03-21
    * Add --csv-eol-unix

0.20	2019-01-01
    * Versions
    * It's 2017
    * Show a ranked summary of the tries with --try
    * It's 2018
    * Deal with proxy settings in env
    * Add index to list/ping lists
    * It's 2019

0.19	2016-01-02
    * Try multiple addresses for the server list RT#110400
    * It's 2016

0.18	2015-10-10
    * Add --prtg option
    * Add --bytes option
    * Cleanup code to calculate average

0.17	2015-03-27
    * Add graph example
    * Update version in .pm

0.16	2015-02-11
    * Set a limit of 40 of the number of hosts to ping
    * Remove perl recommendation from META as it breaks cpan clients

0.15	2015-01-15
    * Renamed to speedtest
    * Do not report/show origin on --list & --ping
    * Documentation/Manual + --man and --info
    * Put on CPAN

0.12	2015-01-24
    * Add --csv
    * Update README

0.11	2015-01-21
    * Add --timeout
    * Update copyright
    * Warning about speedtest not being ping
    * Data::Peek optional (but still recommended)
    * Fallback to distance in absence of GEO-IP

0.10	2014-12-17
    * Add latency for --mini and --url
    * Optionally show server IP
    * Don't show full path in help
    * Fix typo in help (issue#4)
    * Empty line separation in multi-host tests (issue#7)
    * Make README more github-like
    * Add Debian info for requirements
    * Contribution update
    * Browser-neutral agent-ident

0.09	2014-12-02
    * Improve --help
    * Add --simple (alias for -v0)
    * Implement --mini

0.08	2014-11-30
    * Report latency in ms
    * Shorten timeouts on detecting server speeds

0.07	2014-11-26
    * Allow fetching list off ALL servers
    * Country codes (upper case)
    * Ignore country-code when it does not apply

0.06	2014-11-18
    * Add option --one-line
    * Allow specifying a server-url

0.05	2014-10-14
    * Allow test on multiple servers
    * Put on github

0.04	2014-10-10
    * Require 5.12.0 instead of 5.16.3
    * Options -q and -Q accept count
    * Verbose levels
    * Values for min and max speeds druing test
    * Speed report improvement
    * Use random data in upload
    * Remove ignored servers from list

0.03	2014-10-10
    * First checkin to git


speedtest.pl
#!/pro/bin/perl

# speedtest.pl - test network speed using speedtest.net
# (m)'15 [2015-01-21] Copyright H.M.Brand 2014-2015

use 5.10.0;
use warnings;

my $VERSION = "0.12";

sub usage
{
    my $err = shift and select STDERR;
    (my $p = $0) =~ s{.*/}{};
    print <<"EOH";
usage: $p [ --no-geo | --country=NL ] [ --list | --ping ] [ options ]
       --geo          use Geo location (default true) for closest testserver
       --all          include *all* servers (default only in own country)
    -c --country=IS   use ISO country code for closest test server
    -1 --one-line     show summary in one line
    -C --csv          output in CSV (stamp,id,ping,tests,direction,speed,min,max)

    -l --list         list test servers in chosen country sorted by distance
    -p --ping         list test servers in chosen country sorted by latency
       --url          show server url in list

    -s --server=nnn   use testserver with id nnn
    -t --timeout=nnn  set server timeout to nnn seconds
       --url=sss      use specific server url (do not scan) ext php
       --mini=sss     use specific server url (do not scan) ext from sss
       --download     test download speed (default true)
       --upload       test upload   speed (default true)
    -q --quick[=20]   do a      quick test (only the fastest 20 tests)
    -Q --realquick    do a real quick test (only the fastest 10 tests)
    -T --try[=5]      try all tests on the n fastest servers

    -v --verbose[=1]  set verbosity
       --simple       alias for -v0
       --ip           show IP for server
    -V --version      show version and exit
    -? --help         show this help

  $p --list
  $p --ping --country=BE
  $p
  $p -s 4358
  $p --url=http://ookla.extraip.net
  $p -q --no-download
  $p -Q --no-upload
  
EOH
    exit $err;
    } # usage

use Getopt::Long qw(:config bundling noignorecase);
my $opt_c = "";
my $opt_v = 1;
my $opt_d = 1;
my $opt_u = 1;
my $opt_g = 1;
my $opt_q = 0;
my $opt_T = 1;
GetOptions (
    "help|h|?"		=> sub { usage (0); },
    "V|version!"	=> sub { say $VERSION; exit 0; },
    "v|verbose:2"	=>    \$opt_v,
      "simple!"		=> sub { $opt_v = 0; },

      "all!"		=> \my $opt_a,
    "g|geo!"		=>    \$opt_g,
    "c|cc|country=s"	=>    \$opt_c,
    "1|one-line!"	=> \my $opt_1,
    "C|csv!"		=> \my $opt_C,

    "l|list!"		=> \my $list,
    "p|ping!"		=> \my $ping,
      "url:s"		=> \my $url,
      "ip!"		=> \my $ip,

    "T|try:5"		=>    \$opt_T,
    "s|server=i"	=> \my $server,
    "t|timeout=i"	=> \my $timeout,
    "d|download!"	=>    \$opt_d,
    "u|upload!"		=>    \$opt_u,
    "q|quick|fast:20"	=>    \$opt_q,
    "Q|realquick:10"	=>    \$opt_q,

    "m|mini=s"		=> \my $mini,
      "source=s"	=> \my $source,	# NYI
    ) or usage (1);

$opt_C and $opt_v = 0;

use LWP::UserAgent;
use XML::Simple;
use HTML::TreeBuilder;
use Time::HiRes qw( gettimeofday tv_interval );
use Math::Trig;
use Socket qw( inet_ntoa );

# Debugging. Prefer Data::Peek over Data::Dumper if available
{   use Data::Dumper;
    my $dp = 0;
    eval q{
	use Data::Peek;
	$dp = 1;
	};
    sub ddumper
    {
	$dp ? DDumper (@_)
	    : print STDERR Dumper (@_);
	} # ddumper
    }

$timeout ||= 10;
my $ua = LWP::UserAgent->new (
    max_redirect => 2,
    agent        => "speedtest/$VERSION",
    parse_head   => 0,
    timeout      => $timeout,
    cookie_jar   => {},
    );

binmode STDOUT, ":encoding(utf-8)";

# Speedtest.net defines Mbit/s and kbit/s using 1000 as multiplier,
# https://support.speedtest.net/entries/21057567-What-do-mbps-and-kbps-mean-
my $k = 1000;

my $config = get_config ();
my $client = $config->{"client"}   or die "Config saw no client\n";
my $times  = $config->{"times"}    or die "Config saw no times\n";
my $downld = $config->{"download"} or die "Config saw no download\n";
my $upld   = $config->{"upload"}   or die "Config saw no upload\n";
$opt_v > 3 and ddumper {
    client => $client,
    times  => $times,
    down   => $downld,
    up     => $upld,
    };

if ($url || $mini) {
    $opt_g   = 0;
    $opt_c   = "";
    $server  = "";
    my $ping    = 0.05;
    my $name    = "";
    my $sponsor = "CLI";
    if ($mini) {
	my $t0  = [ gettimeofday ];
	my $rsp = $ua->request (HTTP::Request->new (GET => $mini));
	$ping   = tv_interval ($t0);
	$rsp->is_success or die $rsp->status_line . "\n";
	my $tree = HTML::TreeBuilder->new ();
	$tree->parse_content ($rsp->content) or die "Cannot parse\n";
	my $ext = "";
	for ($tree->look_down (_tag => "script")) {
	    my $c = ($_->content)[0] or next;
	    ref $c eq "ARRAY" && $c->[0] &&
		$c->[0] =~ m/\b (?: upload_? | config ) Extension
			     \s*: \s* "? ([^"\s]+) /xi or next;
	    $ext = $1;
	    last;
	    }
	$ext or die "No ext found\n";
	($url = $mini) =~ s{/*$}{/speedtest/upload.$ext};
	$sponsor = $_->as_text for $tree->look_down (_tag => "title");
	$name  ||= $_->as_text for $tree->look_down (_tag => "h1");
	$name  ||= "Speedtest mini";
	}
    else {
	$name = "Local";
	$url =~ m{/\w+\.\w+$} or $url =~ s{/?$}{/speedtest/upload.php};
	my $t0  = [ gettimeofday ];
	my $rsp = $ua->request (HTTP::Request->new (GET => $url));
	$ping   = tv_interval ($t0);
	$rsp->is_success or die $rsp->status_line . "\n";
	}
    (my $host = $url) =~ s{^\w+://([^/]+)(?:/.*)?}{$1};
    $url = {
	cc      => "",
	country => "",
	dist    => "0.0",
	host    => $host,
	id      => 0,
	lat     => "0.0000",
	lon     => "0.0000",
	name    => $name,
	ping    => $ping * 1000,
	sponsor => $sponsor,
	url     => $url,
	url2    => $url,
	};
    }

if ($server) {
    $opt_c = "";
    $opt_a = 1;
    }
else {
    if ($opt_c) {
	$opt_c = uc $opt_c;
	}
    elsif ($opt_g) {	# Try GeoIP
	$opt_v > 5 and say STDERR "Testing Geo location";
	my $url = "http://www.geoiptool.com";
	my $rsp = $ua->request (HTTP::Request->new (GET => $url));
	if ($rsp->is_success) {
	    my $tree = HTML::TreeBuilder->new ();
	    if ($tree->parse_content ($rsp->content)) {
		foreach my $e ($tree->look_down (_tag => "div", class => "data-item")) {
		    $opt_v > 2 and say STDERR $e->as_text;
		    $e->as_text =~ m{Country code(?:\s*:)?\s*([A-Za-z]+)}i or next;
		    $opt_c = uc $1;
		    last;
		    }
		}
	    }
	unless ($opt_c) {	# GEO-Ip failed :/
	    $opt_v and warn "GEO-IP failed. Getting country code based on nearest server\n";
	    my $keep_a = $opt_a;
	    $opt_a = 1;
	    my %list = servers ();
	    my $nearest = { dist => 9999999 };
	    foreach my $id (keys %list) {
		$list{$id}{dist} < $nearest->{dist} and $nearest = $list{$id};
		}
	    $opt_v > 3 and ddumper { nearest => $nearest };
	    $opt_c = $nearest->{cc};
	    $opt_a = $keep_a;
	    }
	}
    $opt_c ||= "IS";	# Iceland seems like a nice default :P
    }

$opt_v and say STDERR "Testing for $client->{ip} : $client->{isp} ($opt_c)";

if ($list) {
    my %list = servers ();
    my @fld = qw( id sponsor name dist );
    my $fmt = "%5d: %-30.30s %-15.15s %7.2f km\n";
    if (defined $url) {
	push @fld, "url0";
	$fmt .= "       %s\n";
	}
    printf $fmt, @{$list{$_}}{@fld}
	for sort { $list{$a}{dist} <=> $list{$b}{dist} } keys %list;
    exit 0;
    }

if ($ping) {
    my @fld = qw( id sponsor name dist ping );
    my $fmt = "%5d: %-30.30s %-15.15s %7.2f km %7.0f ms\n";
    if (defined $url) {
	push @fld, "url0";
	$fmt .= "       %s\n";
	}
    printf $fmt, @{$_}{@fld} for servers_by_ping ();
    exit 0;
    }

# default action is to run on fastest server
my @srvrs = $url ? ($url) : servers_by_ping ();
my @hosts = grep { $_->{ping} < 1000 } @srvrs;
@hosts > $opt_T and splice @hosts, $opt_T;
foreach my $host (@hosts) {
    $host->{sponsor} =~ s/\s+$//;
    if ($opt_v) {
	my $s = "";
	if ($ip) {
	    (my $h =  $host->{url}) =~ s{^\w+://([^/]+)(?:/.*)?$}{$1};
	    my @ad = gethostbyname ($h);
	    $s = join " " => "", map { inet_ntoa ($_) } @ad[4 .. $#ad];
	    }
	@hosts > 1 and print STDERR "\n";
	printf STDERR "Using %4d: %6.2f km %7.0f ms%s %s\n",
	    $host->{id}, $host->{dist}, $host->{ping}, $s, $host->{sponsor};
	}
    $opt_v > 3 and ddumper $host;
    (my $base = $host->{url}) =~ s{/[^/]+$}{};

    my $dl = "-";
    if ($opt_d) {
	$opt_v and print STDERR "Test download ";
	# http://ookla.extraip.net/speedtest/random350x350.jpg
	my @url = map { ("$base/random${_}x${_}.jpg") x 4 }
	    350, 500, 750, 1000, 1500, 2000, 2500, 3000, 3500, 4000;
	my @mnmx = (999999999.999, 0.000);
	my $size = 0;
	my $time = 0;
	$opt_q and splice @url, $opt_q;
	foreach my $url (@url) {
	    my $req = HTTP::Request->new (GET => $url);
	    my $t0 = [ gettimeofday ];
	    my $rsp = $ua->request ($req);
	    my $elapsed = tv_interval ($t0);
	    unless ($rsp->is_success) {
		warn "$url: ", $rsp->status_line, "\n";
		next;
		}
	    my $sz = length $rsp->content;
	    $size += $sz;
	    $time += $elapsed;
	    my $speed = 8 * $sz / $elapsed / $k / $k;
	    $speed < $mnmx[0] and $mnmx[0] = $speed;
	    $speed > $mnmx[1] and $mnmx[1] = $speed;
	    $opt_v     and print  STDERR ".";
	    $opt_v > 2 and printf STDERR "%12.3f %s\n", $speed, $url;
	    }
	$dl = sprintf "%8.3f", 8 * ($size / $time) / $k / $k;
	if ($opt_C) {
	    my @d = localtime;
	    # stamp,id,ping,tests,direction,speed)
	    printf qq{"%4d-%02d-%02d %02d:%02d:%02d",%d,%.2f,%d,D,%.2f,%.2f,%.2f\r\n},
		$d[5] + 1900, ++$d[4], @d[3,2,1,0],
		$host->{id}, $host->{ping},
		scalar @url, $dl, @mnmx;
	    }
	else {
	    $opt_q &&  $opt_v and print " " x (40 - $opt_q);
	    $opt_v || !$opt_1 and print "Download: $dl Mbit/s\n";
	    $opt_v > 1 and printf "  Received %10.2f kb in %9.3f s. [%8.3f - %8.3f]\n",
		$size / 1024, $time, @mnmx;
	    }
	}

    my $ul = "-";
    if ($opt_u) {
	$opt_v and print STDERR "Test upload   ";
	my @data = (0 .. 9, "a" .. "Z", "a" .. "z"); # Random pure ASCII data
	my $data = join "" => map { $data[int rand $#data] } 0 .. 4192;
	$data = "content1=".($data x 1024); # Total length just over 4 Mb
	my @mnmx = (999999999.999, 0.000);
	my $size = 0;
	my $time = 0;
	my $url  = $host->{url}; # .php, .asp, .aspx, .jsp
	# see $upld->{mintestsize} and $upld->{maxchunksize} ?
	my @size = map { $_ * 1000 }
	    ((256) x 10, (512) x 10, (1024) x 10, (4192) x 10);
	$opt_q and splice @size, $opt_q;
	foreach my $sz (@size) {
	    my $req = HTTP::Request->new (POST => $url);
	    $req->content (substr $data, 0, $sz);
	    my $t0 = [ gettimeofday ];
	    my $rsp = $ua->request ($req);
	    my $elapsed = tv_interval ($t0);
	    unless ($rsp->is_success) {
		warn "$url: ", $rsp->status_line, "\n";
		next;
		}
	    $size += $sz;
	    $time += $elapsed;
	    my $speed = 8 * $sz / $elapsed / $k / $k;
	    $speed < $mnmx[0] and $mnmx[0] = $speed;
	    $speed > $mnmx[1] and $mnmx[1] = $speed;
	    $opt_v     and print  STDERR ".";
	    $opt_v > 2 and printf STDERR "%12.3f %s (%7d)\n", $speed, $url, $sz;
	    }
	$ul = sprintf "%8.3f", 8 * ($size / ($time || 1)) / $k / $k;
	if ($opt_C) {
	    my @d = localtime;
	    # stamp,id,ping,tests,direction,speed)
	    printf qq{"%4d-%02d-%02d %02d:%02d:%02d",%d,%.2f,%d,U,%.2f,%.2f,%.2f\r\n},
		$d[5] + 1900, ++$d[4], @d[3,2,1,0],
		$host->{id}, $host->{ping},
		scalar @size, $ul, @mnmx;
	    }
	else {
	    $opt_q &&  $opt_v and print " " x (40 - $opt_q);
	    $opt_v || !$opt_1 and print "Upload:   $ul Mbit/s\n";
	    $opt_v > 1 and printf "  Sent     %10.2f kb in %9.3f s. [%8.3f - %8.3f]\n",
		$size / 1024, $time, @mnmx;
	    }
	}
    $opt_1 and print "DL: $dl Mbit/s, UL: $ul Mbit/s\n";
    }

### ############################################################################

sub get_config
{
    my $url = "http://www.speedtest.net/speedtest-config.php";
    my $rsp = $ua->request (HTTP::Request->new (GET => $url));
    $rsp->is_success or die "Cannot get config: ", $rsp->status_line, "\n";
    my $xml = XMLin ( $rsp->content,
        keeproot        => 1,
        noattr          => 0,
        keyattr         => [ ],
        suppressempty   => "",
        );
    $opt_v > 5 and ddumper $xml->{settings};
    return $xml->{settings};
    } # get_config

sub get_servers
{
#   my $url = "http://www.speedtest.net/speedtest-servers.php";
    my $url = "http://www.speedtest.net/speedtest-servers-static.php";
    my $rsp = $ua->request (HTTP::Request->new (GET => $url));
    $rsp->is_success or die "Cannot get config: ", $rsp->status_line, "\n";
    my $xml = XMLin ( $rsp->content,
        keeproot        => 1,
        noattr          => 0,
        keyattr         => [ ],
        suppressempty   => "",
        );
    # 4601 => {
    #   cc      => 'NL',
    #   country => 'Netherlands',
    #   dist    => '38.5028663935342602',	# added later
    #   id      => 4601,
    #   lat     => '52.2167',
    #   lon     => '5.9667',
    #   name    => 'Apeldoorn',
    #   sponsor => 'Solcon Internetdiensten N.V.',
    #   url     => 'http://speedtest.solcon.net/speedtest/upload.php',
    #   url2    => 'http://ooklaspeedtest.solcon.net/speedtest/upload.php'
    #   },

    return map { $_->{id} => $_ } @{$xml->{settings}{servers}{server}};
    } # get_servers

sub distance
{
    my ($lat_c, $lon_c, $lat_s, $lon_s) = @_;
    my $rad = 6371; # km

    # Convert angles from degrees to radians
    my $dlat = deg2rad ($lat_s - $lat_c);
    my $dlon = deg2rad ($lon_s - $lon_c);

    my $x = sin ($dlat / 2) * sin ($dlat / 2) +
	    cos (deg2rad ($lat_c)) * cos (deg2rad ($lat_s)) *
		sin ($dlon / 2) * sin ($dlon / 2);

    return 6371 * 2 * atan2 (sqrt ($x), sqrt (1 - $x)); # km
    } # distance

sub servers
{
    my %list = get_servers ();
    if (my $iid = $config->{"server-config"}{ignoreids}) {
	$opt_v > 3 and warn "Removing servers $iid from server list\n";
	delete @list{split m/\s*,\s*/ => $iid};
	}
    $opt_a or delete @list{grep { $list{$_}{cc} ne $opt_c } keys %list};
    %list or die "No servers in $opt_c found\n";
    for (values %list) {
	$_->{dist} = distance ($client->{lat}, $client->{lon},
	    $_->{lat}, $_->{lon});
	($_->{url0} = $_->{url}) =~ s{/speedtest/upload.*}{};
	$opt_v > 7 and ddumper $_;
	}
    return %list;
    } # servers

sub servers_by_ping
{
    my %list = servers;
    $opt_v > 1 and say STDERR "Finding fastest host out of @{[scalar keys %list]} hosts for $opt_c ...";
    my $pa = LWP::UserAgent->new (
	max_redirect => 2,
	agent        => "Opera/25.00 opera 25",
	parse_head   => 0,
	cookie_jar   => {},
	timeout      => $timeout,
	);
    foreach my $h (values %list) {
	my $t = 0;
	if ($server and $h->{id} != $server) {
	    $h->{ping} = 40000;
	    next;
	    }
	$opt_v > 5 and printf STDERR "? %4d %-20.20s %s\n",
	    $h->{id}, $h->{sponsor}, $h->{url};
	my $req = HTTP::Request->new (GET => "$h->{url}/latency.txt");
	for (0 .. 3) {
	    my $t0 = [ gettimeofday ];
	    my $rsp = $pa->request ($req);
	    my $elapsed = tv_interval ($t0);
	    $opt_v > 8 and printf STDERR "%4d %9.2f\n", $_, $elapsed;
	    if ($elapsed >= 15) {
		$t = 40;
		last;
		}
	    $t += ($rsp->is_success ? $elapsed : 1000);
	    }
	$h->{ping} = $t * 1000; # report in ms
	}
    return sort { $a->{ping} <=> $b->{ping}
	       || $a->{dist} <=> $b->{dist} } values %list;
    } # servers_by_ping
